File: src/main/java/io/github/funofprograming/context/impl/ConcurrentApplicationContextImpl.java
package io.github.funofprograming.context.impl;

import java.util.ConcurrentModificationException;
import java.util.Set;

import io.github.funofprograming.context.ApplicationContext;
import io.github.funofprograming.context.Key;
import io.github.funofprograming.context.ApplicationContextMergeStrategy;
import io.github.funofprograming.context.ConcurrentApplicationContext;
import lombok.AccessLevel;
import lombok.Getter;

/**
 * Thread safe extension of {@linkplain ApplicationContextImpl} by implementing {@linkplain ConcurrentApplicationContextImpl}
 * 
 * @author Akshay Jain
 *
 */
public class ConcurrentApplicationContextImpl extends ApplicationContextImpl implements ConcurrentApplicationContext
{
    @Getter(AccessLevel.PROTECTED)
    private final ContextConcurrencyManager contextConcurrencyManager;

    /**
     * {@inheritDoc}
     */
    public ConcurrentApplicationContextImpl(String name) 
    {
        this(name, null);
    }
    
    /**
     * {@inheritDoc}
     */
    public ConcurrentApplicationContextImpl(String name, Set<Key<?>> permittedKeys) 
    {
        super(name, permittedKeys);
        this.contextConcurrencyManager = ContextConcurrencyManager.getInstance();
    }
    
    /**
     * {@inheritDoc}
     */
    @Override
    public <T> void addIfNotPresent(Key<T> key, T value)
    {
        if(!getContextConcurrencyManager().acquireWriteLock())
        {
            throw new ConcurrentModificationException("Some other thread modifying this context at the moment");
        }
        
        try
        {
            super.addIfNotPresent(key, value);
        } 
        finally
        {
            getContextConcurrencyManager().releaseWriteLock();
        }
    }
    
    /**
     * {@inheritDoc}
     */
    @Override
    public <T> void addIfNotPresent(Key<T> key, T value, Long timeout) 
    {
        if(!getContextConcurrencyManager().acquireWriteLock(timeout))
        {
            throw new ConcurrentModificationException("Some other thread modifying this context at the moment");
        }
        
        try
        {
            super.addIfNotPresent(key, value);
        } 
        finally
        {
            getContextConcurrencyManager().releaseWriteLock();
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public <T> T addWithOverwrite(Key<T> key, T value) 
    {
        if(!getContextConcurrencyManager().acquireWriteLock())
        {
            throw new ConcurrentModificationException("Some other thread modifying this context at the moment");
        }
        
        try
        {
            return super.addWithOverwrite(key, value);
        }
        finally
        {
            getContextConcurrencyManager().releaseWriteLock();
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public <T> T addWithOverwrite(Key<T> key, T value, Long timeout) 
    {
        if(!getContextConcurrencyManager().acquireWriteLock(timeout))
        {
            throw new ConcurrentModificationException("Some other thread modifying this context at the moment");
        }
        
        try
        {
            return super.addWithOverwrite(key, value);
        }
        finally
        {
            getContextConcurrencyManager().releaseWriteLock();
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public <T> void add(Key<T> key, T value)
    {
        if(!getContextConcurrencyManager().acquireWriteLock())
        {
            throw new ConcurrentModificationException("Some other thread modifying this context at the moment");
        }
        
        try
        {
            super.add(key, value);
        }
        finally
        {
            getContextConcurrencyManager().releaseWriteLock();
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public <T> void add(Key<T> key, T value, Long timeout) 
    {
        if(!getContextConcurrencyManager().acquireWriteLock(timeout))
        {
            throw new ConcurrentModificationException("Some other thread modifying this context at the moment");
        }
        
        try
        {
            super.add(key, value);
        }
        finally
        {
            getContextConcurrencyManager().releaseWriteLock();
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public <T> boolean exists(Key<T> key) 
    {
        if(!getContextConcurrencyManager().acquireReadLock())
        {
            throw new ConcurrentModificationException("Some other thread modifying this context at the moment");
        }
        
        try
        {
            return super.exists(key);
        }
        finally
        {
            getContextConcurrencyManager().releaseReadLock();
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public <T> boolean exists(Key<T> key, Long timeout) 
    {
        if(!getContextConcurrencyManager().acquireReadLock(timeout))
        {
            throw new ConcurrentModificationException("Some other thread modifying this context at the moment");
        }
        
        try
        {
            return super.exists(key);
        }
        finally
        {
            getContextConcurrencyManager().releaseReadLock();
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public <T> T fetch(Key<T> key) 
    {
        if(!getContextConcurrencyManager().acquireReadLock())
        {
            throw new ConcurrentModificationException("Some other thread modifying this context at the moment");
        }
        
        try
        {
            return super.fetch(key);
        }
        finally
        {
            getContextConcurrencyManager().releaseReadLock();
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public <T> T fetch(Key<T> key, Long timeout) 
    {
        if(!getContextConcurrencyManager().acquireReadLock(timeout))
        {
            throw new ConcurrentModificationException("Some other thread modifying this context at the moment");
        }
        
        try
        {
            return super.fetch(key);
        }
        finally
        {
            getContextConcurrencyManager().releaseReadLock();
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public <T> T erase(Key<T> key) 
    {
        if(!getContextConcurrencyManager().acquireWriteLock())
        {
            throw new ConcurrentModificationException("Some other thread modifying this context at the moment");
        }
        
        try
        {
            return super.erase(key);
        }
        finally
        {
            getContextConcurrencyManager().releaseWriteLock();
        }
    }
    
    /**
     * {@inheritDoc}
     */
    @Override
    public <T> T erase(Key<T> key, Long timeout) 
    {
        if(!getContextConcurrencyManager().acquireWriteLock(timeout))
        {
            throw new ConcurrentModificationException("Some other thread modifying this context at the moment");
        }
        
        try
        {
            return super.erase(key);
        }
        finally
        {
            getContextConcurrencyManager().releaseWriteLock();
        }
    }
    
    /**
     * {@inheritDoc}
     */
    @Override
    public void clear() 
    {
        if(!getContextConcurrencyManager().acquireWriteLock())
        {
            throw new ConcurrentModificationException("Some other thread modifying this context at the moment");
        }
        
        try
        {
            super.clear();
        }
        finally
        {
            getContextConcurrencyManager().releaseWriteLock();
        }
    }
    
    /**
     * {@inheritDoc}
     */
    @Override
    public void clear(Long timeout) 
    {
        if(!getContextConcurrencyManager().acquireWriteLock(timeout))
        {
            throw new ConcurrentModificationException("Some other thread modifying this context at the moment");
        }
        
        try
        {
            super.clear();
        }
        finally
        {
            getContextConcurrencyManager().releaseWriteLock();
        }
    }
    
    @Override
    public void merge(ApplicationContext other, ApplicationContextMergeStrategy mergeStrategy, Long timeout)
    {
        if(!getContextConcurrencyManager().acquireWriteLock(timeout))
        {
            throw new ConcurrentModificationException("Some other thread modifying this context at the moment");
        }
        
        try
        {
            super.merge(other, mergeStrategy);
        }
        finally
        {
            getContextConcurrencyManager().releaseWriteLock();
        }
    }

    @Override
    public void merge(ApplicationContext other, ApplicationContextMergeStrategy mergeStrategy)
    {
        this.merge(other, mergeStrategy, null);
    }
}

File: src/main/kotlin/io/github/funofprograming/context/impl/ConcurrentApplicationContextImpl.kt
package io.github.funofprograming.context.impl

import io.github.funofprograming.context.ApplicationContext
import io.github.funofprograming.context.ApplicationContextMergeStrategy
import io.github.funofprograming.context.ConcurrentApplicationContext
import io.github.funofprograming.context.Key
import kotlinx.coroutines.runBlocking

open class ConcurrentApplicationContextImpl(private val name: String, private val permittedKeys: Set<Key<*>>? = null) : ApplicationContextImpl(name, permittedKeys), ConcurrentApplicationContext {

    private val concurrencyManager = ContextConcurrencyManager.getInstance()

    override fun <T> addIfNotPresent(key: Key<T>, value: T?) = runBlocking { addIfNotPresent(key, value, null) }

    override suspend fun <T> addIfNotPresent(key: Key<T>, value: T?, timeout: Long?) = concurrencyManager.executeWriteWithLock(timeout) { super.addIfNotPresent(key, value) }

    override fun <T> addWithOverwrite(key: Key<T>, value: T?): Any? = runBlocking { addWithOverwrite(key, value, null) }

    override suspend fun <T> addWithOverwrite(key: Key<T>, value: T?, timeout: Long?): Any? = concurrencyManager.executeWriteWithLock(timeout) { super.addWithOverwrite(key, value) }

    override fun <T> add(key: Key<T>, value: T?) = runBlocking { add(key, value, null) }

    override suspend fun <T> add(key: Key<T>, value: T?, timeout: Long?) = concurrencyManager.executeWriteWithLock(timeout) { super.add(key, value) }

    override fun <T> exists(key: Key<T>): Boolean = runBlocking { exists(key, null) }

    override suspend fun <T> exists(key: Key<T>, timeout: Long?): Boolean = concurrencyManager.executeReadWithLock(timeout) { super.exists(key) }

    override fun <T> fetch(key: Key<T>): T? = runBlocking { fetch(key, null) }

    override suspend fun <T> fetch(key: Key<T>, timeout: Long?): T? = concurrencyManager.executeReadWithLock(timeout) { super.fetch(key) }

    override fun <T> erase(key: Key<T>): T? = runBlocking { erase(key, null) }

    override suspend fun <T> erase(key: Key<T>, timeout: Long?): T? = concurrencyManager.executeWriteWithLock(timeout) { super.erase(key) }

    override fun clear() = runBlocking { clear(null) }

    override suspend fun clear(timeout: Long?) = concurrencyManager.executeWriteWithLock(timeout) { super.clear() }

    override fun merge(other: ApplicationContext, mergeStrategy: ApplicationContextMergeStrategy) = runBlocking { merge(other, mergeStrategy, null) }

    override suspend fun merge(other: ApplicationContext, mergeStrategy: ApplicationContextMergeStrategy, timeout: Long?) = concurrencyManager.executeWriteWithLock(timeout) { super.merge(other, mergeStrategy) }
}